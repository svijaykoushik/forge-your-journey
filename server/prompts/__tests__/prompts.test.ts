import {
  generateAdventureOutlinePrompt,
  generateWorldDetailsPrompt,
  generateStorySegmentPrompt,
  generateActionFeasibilityPrompt,
  generateCustomActionOutcomePrompt,
  generateAttemptToFixJsonPrompt,
  generateSceneExaminationPrompt,
  genrePersonaDetails
} from '../prompts'; // Removed .js extension
import { AdventureOutline, Persona, StorySegment, WorldDetails, GameGenre, InventoryItem } from '../../../types'; // Removed .js extension

describe('server/prompts.ts', () => {
  // Mock data for testing prompts
  const mockGenre: GameGenre = 'Dark Fantasy';
  const mockPersona: Persona = 'Brave Warrior';
  const mockAdventureOutline: AdventureOutline = {
    title: 'The Sunless Citadel',
    overallGoal: 'Reclaim the Sunstone from the heart of the citadel.',
    stages: [
      { title: 'Entry', description: 'Breach the citadel gates.', objective: 'Find a way in.' },
      { title: 'The Undercroft', description: 'Navigate the haunted halls.', objective: 'Reach the inner sanctum.' },
      { title: 'Sanctum of Shadows', description: 'Confront the guardian and claim the Sunstone.', objective: 'Defeat the guardian.' },
    ],
  };
  const mockWorldDetails: WorldDetails = {
    worldName: 'Aethel',
    genreClarification: 'Gritty low-magic fantasy',
    keyEnvironmentalFeatures: ['A perpetual twilight covers the land', 'Twisted, thorny forests'],
    dominantSocietiesOrFactions: ['The Iron Concord (human supremacists)', 'The Silent Hand (thieves guild)'],
    uniqueCreaturesOrMonsters: ['Gloomfang Wyrm', 'Shadow Stalkers'],
    magicSystemOverview: 'Magic is rare and dangerous, often corrupting its users.',
    briefHistoryHook: 'A great war centuries ago shattered the old kingdoms, leaving behind ruins and bitterness.',
    culturalNormsOrTaboos: ['Never trust a sorcerer', 'Obsidian is a symbol of bad luck'],
  };
  const mockCurrentSegment: StorySegment = {
    sceneDescription: 'You stand before a crumbling fortress, an eerie silence blanketing the air.',
    choices: [],
    imagePrompt: 'A crumbling fortress under a dark sky.',
    isUserInputCommandOnly: true,
  };
  const mockInventory: InventoryItem[] = [
    { id: 'torch', name: 'Torch', description: 'A flickering torch to light your way.' },
  ];

  describe('generateAdventureOutlinePrompt', () => {
    it('should generate the correct prompt for adventure outline', () => {
      const prompt = generateAdventureOutlinePrompt(mockGenre, mockPersona);
      expect(prompt).toContain(mockGenre);
      expect(prompt).toContain(genrePersonaDetails[mockGenre]?.[mockPersona]?.title || mockPersona);
      expect(prompt).toMatchSnapshot();
    });
  });

  describe('generateWorldDetailsPrompt', () => {
    it('should generate the correct prompt for world details', () => {
      const prompt = generateWorldDetailsPrompt(mockAdventureOutline, mockPersona, mockGenre);
      expect(prompt).toContain(mockAdventureOutline.title);
      expect(prompt).toContain(mockAdventureOutline.overallGoal);
      expect(prompt).toContain(genrePersonaDetails[mockGenre]?.[mockPersona]?.title || mockPersona);
      expect(prompt).toMatchSnapshot();
    });
  });

  describe('generateStorySegmentPrompt', () => {
    const baseFullPrompt = `The player chose to open the ancient chest.`;
    it('should generate the correct prompt for an initial story segment', () => {
      const prompt = generateStorySegmentPrompt(baseFullPrompt, true);
      expect(prompt).toContain(baseFullPrompt);
      expect(prompt).toContain('For this initial scene');
      expect(prompt).toMatchSnapshot();
    });
    it('should generate the correct prompt for a subsequent story segment', () => {
      const prompt = generateStorySegmentPrompt(baseFullPrompt, false);
      expect(prompt).toContain(baseFullPrompt);
      expect(prompt).toContain('For subsequent scenes');
      expect(prompt).toMatchSnapshot();
    });
  });

  describe('generateActionFeasibilityPrompt', () => {
    it('should generate the correct prompt for action feasibility', () => {
      const userInput = 'I try to scale the fortress wall.';
      const prompt = generateActionFeasibilityPrompt(
        userInput,
        mockCurrentSegment,
        mockAdventureOutline,
        mockWorldDetails,
        mockGenre,
        mockPersona,
        mockInventory,
        0 // currentStageIndex
      );
      expect(prompt).toContain(userInput);
      expect(prompt).toContain(mockCurrentSegment.sceneDescription);
      expect(prompt).toContain(mockAdventureOutline.title);
      expect(prompt).toContain(mockWorldDetails.worldName);
      expect(prompt).toContain(mockInventory[0].name);
      expect(prompt).toMatchSnapshot();
    });
  });

  describe('generateCustomActionOutcomePrompt', () => {
    const userInput = 'I use my torch to light the way.';
    const feasibilityContextPossible = {
        wasImpossible: false,
        suggestionIfPossible: "The torch reveals a hidden inscription."
    };
    const feasibilityContextImpossible = {
        wasImpossible: true,
        reasonForImpossibility: "The torch is too damp to light."
    };

    it('should generate the correct prompt for a possible custom action outcome', () => {
      const prompt = generateCustomActionOutcomePrompt(
        userInput,
        mockCurrentSegment,
        mockAdventureOutline,
        mockWorldDetails,
        mockGenre,
        mockPersona,
        mockInventory,
        0, // currentStageIndex
        feasibilityContextPossible
      );
      expect(prompt).toContain(userInput);
      expect(prompt).toContain(feasibilityContextPossible.suggestionIfPossible);
      expect(prompt).toMatchSnapshot();
    });

    it('should generate the correct prompt for an impossible custom action outcome', () => {
      const prompt = generateCustomActionOutcomePrompt(
        userInput,
        mockCurrentSegment,
        mockAdventureOutline,
        mockWorldDetails,
        mockGenre,
        mockPersona,
        mockInventory,
        0, // currentStageIndex
        feasibilityContextImpossible
      );
      expect(prompt).toContain(userInput);
      expect(prompt).toContain(feasibilityContextImpossible.reasonForImpossibility);
      expect(prompt).toMatchSnapshot();
    });
  });

  describe('generateAttemptToFixJsonPrompt', () => {
    it('should generate the correct prompt for attempting to fix JSON', () => {
      const faultyJson = '{"name": "Test", "description": "A test object",'; // Missing closing brace
      const originalContext = 'User was trying to create an item.';
      const prompt = generateAttemptToFixJsonPrompt(faultyJson, originalContext);
      expect(prompt).toContain(faultyJson);
      expect(prompt).toContain(originalContext.substring(0,500));
      expect(prompt).toMatchSnapshot();
    });
  });

  describe('generateSceneExaminationPrompt', () => {
    it('should generate the correct prompt for scene examination', () => {
      const prompt = generateSceneExaminationPrompt(
        mockCurrentSegment.sceneDescription,
        mockGenre,
        mockAdventureOutline,
        mockWorldDetails,
        mockAdventureOutline.stages[0].title,
        mockAdventureOutline.stages[0].objective,
        mockPersona,
        mockInventory
      );
      expect(prompt).toContain(mockCurrentSegment.sceneDescription);
      expect(prompt).toContain(mockWorldDetails.worldName);
      expect(prompt).toContain(mockInventory[0].name);
      expect(prompt).toMatchSnapshot();
    });
  });
});
