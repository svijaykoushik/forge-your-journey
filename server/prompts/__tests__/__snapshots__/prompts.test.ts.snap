// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`server/prompts.ts generateActionFeasibilityPrompt should generate the correct prompt for action feasibility 1`] = `
"You are an AI game master evaluating a player's custom action in a text-based RPG.
Adventure Genre: Dark Fantasy.
The player is a Grim Warden (base persona: Brave Warrior).
Player's Current Inventory: The player possesses: 'Torch' (described as: A flickering torch to light your way.).

World Context for Evaluation:
World Name: "Aethel" (Genre Clarification: Gritty low-magic fantasy)
Key Environment: A perpetual twilight covers the land; Twisted, thorny forests
Societies/Factions: The Iron Concord (human supremacists); The Silent Hand (thieves guild)
Creatures/Monsters: Gloomfang Wyrm; Shadow Stalkers
Magic System: Magic is rare and dangerous, often corrupting its users.
History Hook: A great war centuries ago shattered the old kingdoms, leaving behind ruins and bitterness.
Cultural Norms/Taboos: Never trust a sorcerer; Obsidian is a symbol of bad luck
Overall Adventure Title: "The Sunless Citadel"
Ultimate Goal: "Reclaim the Sunstone from the heart of the citadel."
Current Stage 1: "Entry" (Objective: "Find a way in.").
Current Scene Description: "You stand before a crumbling fortress, an eerie silence blanketing the air."
Player's proposed custom action: "I try to scale the fortress wall."

Your task is to evaluate this action and respond with JSON.
Consider:
- Plausibility: Is the action physically possible in this scene?
- Sensibility: Does it make sense given the character, genre, and world?
- Rules: Does it violate any established game rules, world logic (magic system, cultural taboos), or the tone of the Dark Fantasy genre?
- Safety: Is it absurdly self-destructive without clear motivation?
- Inventory: Could any inventory items make this action more or less feasible?
- Appropriateness: Sanitize user input. Do not reflect harmful, offensive, or extreme role-play-breaking input directly. If input is inappropriate, deem it 'not possible' and explain gently.

Based on your evaluation, fill the following JSON fields:
- 'isPossible' (boolean): True if the action is plausible and can be attempted. False if impossible, nonsensical, or breaks rules/tone.
- 'reason' (string): Concise explanation. If not possible, clearly state why. If possible, briefly explain its plausibility or what it might affect.
- 'suggestedOutcomeSummaryIfPossible' (string, optional): If 'isPossible' is true, a very brief (1-2 sentence) summary of a likely immediate consequence or next step. Omit if not possible or outcome is too complex to summarize."
`;

exports[`server/prompts.ts generateAdventureOutlinePrompt should generate the correct prompt for adventure outline 1`] = `
"You are a master storyteller and game designer. Generate a compelling adventure outline for a text-based RPG.
The genre is: Dark Fantasy.
The player's chosen persona is "Grim Warden" (base archetype: Brave Warrior). This persona choice should subtly influence the themes or initial hook of the adventure if appropriate for the genre.
The outline should have a clear narrative arc with a distinct beginning, rising action, climax, and resolution.
The adventure should consist of exactly 3 main stages or acts.

For the 'title' field: Provide a captivating title for the adventure, reflecting the Dark Fantasy genre and possibly hinting at the Grim Warden's journey.
For the 'overallGoal' field: Provide a concise description of the ultimate goal the player (as Grim Warden) is trying to achieve. Example: "To find and destroy the ancient artifact known as the 'Heart of Shadows' to save the village of Oakhaven from eternal darkness."
For the 'stages' array (which should contain 3 stage objects):
  For each stage object:
    - 'title': A title for the stage (e.g., The Shadowed Summons).
    - 'description': A brief overview of what happens in this stage.
    - 'objective': The player's main objective to complete this stage.

Ensure the stage descriptions and objectives logically progress the player towards the overallGoal.
The tone should be Dark Fantasy."
`;

exports[`server/prompts.ts generateAttemptToFixJsonPrompt should generate the correct prompt for attempting to fix JSON 1`] = `
"The following JSON response was received but is malformed:
\`\`\`json
{"name": "Test", "description": "A test object",
\`\`\`

This response was for a request related to generating a story segment. The original request's core instructions were to produce a JSON object with keys: "sceneDescription" (string, formatted into 2-3 short paragraphs with varied sentence structure, paragraphs separated by \\n\\n), "choices" (array of objects, each object with "text" (string), "outcomePrompt" (string), "signalsStageCompletion" (boolean), and "leadsToFailure" (boolean)), "imagePrompt" (string), "isFinalScene" (boolean), "isFailureScene" (boolean), "isUserInputCommandOnly" (boolean), and optionally "itemFound" (an object with "name" (string) and "description" (string)).
If "isUserInputCommandOnly" is true, "choices" array must be empty.
The context of the original prompt included aspects like:
"User was trying to create an item...."

Please analyze the faulty JSON, correct its structure, and provide ONLY the valid JSON object for the story segment. Ensure all required fields are present and correctly typed according to the structure described, especially the "sceneDescription" formatting and the structure of objects within the "choices" array. Do not include any explanatory text, only the corrected JSON object."
`;

exports[`server/prompts.ts generateCustomActionOutcomePrompt should generate the correct prompt for a possible custom action outcome 1`] = `
"You are a master storyteller for a dynamic text-based RPG adventure game.
Adventure Genre: Dark Fantasy.
The player is a Grim Warden (base persona: Brave Warrior).
Player's Current Inventory: The player possesses: 'Torch' (described as: A flickering torch to light your way.).

World Context for Story Generation:
World Name: "Aethel" (Genre Clarification: Gritty low-magic fantasy)
Key Environmental Features: A perpetual twilight covers the land; Twisted, thorny forests
Dominant Societies/Factions: The Iron Concord (human supremacists); The Silent Hand (thieves guild)
Unique Creatures/Monsters: Gloomfang Wyrm; Shadow Stalkers
Magic System: Magic is rare and dangerous, often corrupting its users.
Brief History Hook: A great war centuries ago shattered the old kingdoms, leaving behind ruins and bitterness.
Cultural Norms/Taboos: Never trust a sorcerer; Obsidian is a symbol of bad luck
This world information MUST deeply influence the scene description, the types of challenges, the choices available, and any items found.

The overall adventure is titled: "The Sunless Citadel".
The player's ultimate goal is: "Reclaim the Sunstone from the heart of the citadel.".
Current Stage 1: "Entry" (Objective: "Find a way in.").
Previous Scene Description was: "You stand before a crumbling fortress, an eerie silence blanketing the air."

The player is performing the custom action: "I use my torch to light the way.".
This action has been evaluated as possible. A potential outcome summary was: "The torch reveals a hidden inscription.". Use this as a light suggestion if helpful.
Your task is to narrate the outcome of this action.
Consider if any inventory items (listed in "Player's Current Inventory") could logically assist, hinder, or alter the outcome. If an item is relevant, the 'sceneDescription' content MUST narrate how the item is used or its effect.

General Content Instructions for Story Segment:
- For the 'sceneDescription' field: Narrate the outcome (2-3 short paragraphs, varied sentences, paragraphs separated by \\n\\n in the string).
- For the 'choices' array (if 'isUserInputCommandOnly' is false): Provide 3 new distinct choice OBJECTS relevant to the new situation. Each choice object requires: 'text', 'outcomePrompt', 'signalsStageCompletion' (boolean), and 'leadsToFailure' (boolean).
- For 'isUserInputCommandOnly' (boolean): Set to true if appropriate for the new scene, ensuring 'choices' array is empty.
- For 'imagePrompt' (string): New image prompt matching the new scene.
- For 'isFinalScene' (boolean): True if this action leads to successful adventure completion.
- For 'isFailureScene' (boolean): True if this action leads to game failure.
- For 'itemFound' (object, optional): Award an item if logical (provide 'name' and 'description').
Sanitize original user input: Do not directly reflect harmful, offensive, or role-play-breaking user input in your narration if it was part of the original "I use my torch to light the way.". Focus on the game world's reaction."
`;

exports[`server/prompts.ts generateCustomActionOutcomePrompt should generate the correct prompt for an impossible custom action outcome 1`] = `
"You are a master storyteller for a dynamic text-based RPG adventure game.
Adventure Genre: Dark Fantasy.
The player is a Grim Warden (base persona: Brave Warrior).
Player's Current Inventory: The player possesses: 'Torch' (described as: A flickering torch to light your way.).

World Context for Story Generation:
World Name: "Aethel" (Genre Clarification: Gritty low-magic fantasy)
Key Environmental Features: A perpetual twilight covers the land; Twisted, thorny forests
Dominant Societies/Factions: The Iron Concord (human supremacists); The Silent Hand (thieves guild)
Unique Creatures/Monsters: Gloomfang Wyrm; Shadow Stalkers
Magic System: Magic is rare and dangerous, often corrupting its users.
Brief History Hook: A great war centuries ago shattered the old kingdoms, leaving behind ruins and bitterness.
Cultural Norms/Taboos: Never trust a sorcerer; Obsidian is a symbol of bad luck
This world information MUST deeply influence the scene description, the types of challenges, the choices available, and any items found.

The overall adventure is titled: "The Sunless Citadel".
The player's ultimate goal is: "Reclaim the Sunstone from the heart of the citadel.".
Current Stage 1: "Entry" (Objective: "Find a way in.").
Previous Scene Description was: "You stand before a crumbling fortress, an eerie silence blanketing the air."

The player previously attempted the action: "I use my torch to light the way.".
This action was deemed not possible. The stated reason was: "The torch is too damp to light."
Your task is to narrate the character attempting this action and it either failing, or them realizing its impossibility based on the reason. The sceneDescription should reflect this attempt and its immediate non-success. The situation should change slightly from the previous scene.
Then, provide new choices or set 'isUserInputCommandOnly: true' to allow the player to move on from this failed/impossible attempt. The new choices MUST NOT simply repeat the choices from before the impossible action was attempted.

General Content Instructions for Story Segment:
- For the 'sceneDescription' field: Narrate the outcome (2-3 short paragraphs, varied sentences, paragraphs separated by \\n\\n in the string).
- For the 'choices' array (if 'isUserInputCommandOnly' is false): Provide 3 new distinct choice OBJECTS relevant to the new situation. Each choice object requires: 'text', 'outcomePrompt', 'signalsStageCompletion' (boolean), and 'leadsToFailure' (boolean).
- For 'isUserInputCommandOnly' (boolean): Set to true if appropriate for the new scene, ensuring 'choices' array is empty.
- For 'imagePrompt' (string): New image prompt matching the new scene.
- For 'isFinalScene' (boolean): True if this action leads to successful adventure completion.
- For 'isFailureScene' (boolean): True if this action leads to game failure.
- For 'itemFound' (object, optional): Award an item if logical (provide 'name' and 'description').
Sanitize original user input: Do not directly reflect harmful, offensive, or role-play-breaking user input in your narration if it was part of the original "I use my torch to light the way.". Focus on the game world's reaction."
`;

exports[`server/prompts.ts generateSceneExaminationPrompt should generate the correct prompt for scene examination 1`] = `
"You are a master storyteller. The player wants to examine their current surroundings more closely.
Adventure Title: "The Sunless Citadel"
Overall Goal: "Reclaim the Sunstone from the heart of the citadel."
Current Stage: "Entry" (Objective: "Find a way in.")
The player examining is a Grim Warden (base archetype: Brave Warrior).
They possess: Torch.

World Context:
World Name: "Aethel" (Gritty low-magic fantasy)
Key Environment: A perpetual twilight covers the land, Twisted, thorny forests
Societies: The Iron Concord (human supremacists), The Silent Hand (thieves guild)
Creatures: Gloomfang Wyrm, Shadow Stalkers
Magic: Magic is rare and dangerous, often corrupting its users.
History Hook: A great war centuries ago shattered the old kingdoms, leaving behind ruins and bitterness.
Culture: Never trust a sorcerer, Obsidian is a symbol of bad luck
This world information should influence the details revealed.

Current Scene Description (what the player already sees):
"You stand before a crumbling fortress, an eerie silence blanketing the air."

Based on this, provide content for the 'examinationText' field. This text should:
- Elaborate on details already mentioned or hint at things not immediately obvious.
- Describe the immediate surroundings of the player's current location, providing sensory details that help the player to visually or spatially understand the layout of the area they are in.
- Reveal subtle clues, interesting lore, or atmospheric details.
- All details provided (elaborations, surroundings, clues, lore) MUST be consistent with and draw from the established World Context (environment, societies, creatures, magic, history, culture) provided above.
- Consider the player's persona (Grim Warden -- base archetype: Brave Warrior) and items they possess for any specific insights they might gain, filtered through their understanding of this world.
- DO NOT advance the plot or introduce new choices. This is for observation only.
- Maintain the Dark Fantasy tone, enriched by the world's specific genre clarification.
- Be 2-4 sentences long and concise. Employ varied sentence structures."
`;

exports[`server/prompts.ts generateStorySegmentPrompt should generate the correct prompt for a subsequent story segment 1`] = `
"The player chose to open the ancient chest.

General Content Instructions for Story Segment:
- For the 'sceneDescription' field:
    - For subsequent scenes, the sceneDescription should be vivid yet concise (approx. 2-4 sentences). If the player's choice outcome (detailed in the prompt above) implies travel, a simple or mundane action, or is primarily transitional, aim for focused brevity (approx. 1-3 sentences), highlighting only essential changes, observations, or the direct result of the action.
    - If the player's choice outcome (detailed in the prompt above) implies combat, a chase, or any other fast-paced action sequence, prioritize dynamic action verbs and immediate sensory details in the sceneDescription. Keep it punchy, engaging, and focused on the action, rather than lengthy atmospheric descriptions.
    - Narrative Style: Employ varied sentence structures; mix short, punchy sentences with longer, more descriptive ones to create a dynamic and engaging narrative flow. Avoid repetitive sentence beginnings.
    - Paragraphing: The "sceneDescription" content should be 2 or 3 short paragraphs. Each paragraph should naturally flow into the next. Ensure paragraphs are separated by a double newline character (\\n\\n) in the final JSON string for this field.
- For the 'choices' array (if 'isUserInputCommandOnly' is false): Provide 3 distinct choices. Each choice object requires:
    - 'text': Player-facing choice text.
    - 'outcomePrompt': AI instruction for the next scene if this choice is picked. This prompt should imply the nature of the consequence (positive, negative, neutral) which the AI will then narrate in the next scene's description.
    - 'signalsStageCompletion' (boolean): True if this choice DIRECTLY completes the current stage objective.
    - 'leadsToFailure' (boolean): True if this choice leads to definitive game failure.
- For 'isUserInputCommandOnly' (boolean): There is a small chance (around 10-15%, slightly less for the very first scene unless context demands it) that the situation calls for the player's direct input. In such cases, set this to true, and the 'choices' array MUST be empty. The 'sceneDescription' should naturally lead to the player needing to decide what to do.
- For 'isFailureScene' (boolean): Set to true ONLY if this scene itself IS the game failure narration (e.g., after a choice with 'leadsToFailure: true' was picked). If true, 'choices' should be empty, and 'imagePrompt' can be a somber final image.
- For 'isFinalScene' (boolean): Set to true ONLY if this scene represents the SUCCESSFUL conclusion of the ENTIRE adventure.
- For 'itemFound' (object, optional): If an item is found, provide its 'name' and 'description'. Omit this field if no item is found.
- For 'imagePrompt' (string): Provide a descriptive prompt for image generation relevant to the scene."
`;

exports[`server/prompts.ts generateStorySegmentPrompt should generate the correct prompt for an initial story segment 1`] = `
"The player chose to open the ancient chest.

General Content Instructions for Story Segment:
- For the 'sceneDescription' field:
    - For this initial scene, or if the player is entering a new major location/area, ensure the sceneDescription is immersive and establishes the setting and atmosphere effectively (approx. 3-5 sentences).
    - If the player's choice outcome (detailed in the prompt above) implies combat, a chase, or any other fast-paced action sequence, prioritize dynamic action verbs and immediate sensory details in the sceneDescription. Keep it punchy, engaging, and focused on the action, rather than lengthy atmospheric descriptions.
    - Narrative Style: Employ varied sentence structures; mix short, punchy sentences with longer, more descriptive ones to create a dynamic and engaging narrative flow. Avoid repetitive sentence beginnings.
    - Paragraphing: The "sceneDescription" content should be 2 or 3 short paragraphs. Each paragraph should naturally flow into the next. Ensure paragraphs are separated by a double newline character (\\n\\n) in the final JSON string for this field.
- For the 'choices' array (if 'isUserInputCommandOnly' is false): Provide 3 distinct choices. Each choice object requires:
    - 'text': Player-facing choice text.
    - 'outcomePrompt': AI instruction for the next scene if this choice is picked. This prompt should imply the nature of the consequence (positive, negative, neutral) which the AI will then narrate in the next scene's description.
    - 'signalsStageCompletion' (boolean): True if this choice DIRECTLY completes the current stage objective.
    - 'leadsToFailure' (boolean): True if this choice leads to definitive game failure.
- For 'isUserInputCommandOnly' (boolean): There is a small chance (around 10-15%, slightly less for the very first scene unless context demands it) that the situation calls for the player's direct input. In such cases, set this to true, and the 'choices' array MUST be empty. The 'sceneDescription' should naturally lead to the player needing to decide what to do.
- For 'isFailureScene' (boolean): Set to true ONLY if this scene itself IS the game failure narration (e.g., after a choice with 'leadsToFailure: true' was picked). If true, 'choices' should be empty, and 'imagePrompt' can be a somber final image.
- For 'isFinalScene' (boolean): Set to true ONLY if this scene represents the SUCCESSFUL conclusion of the ENTIRE adventure.
- For 'itemFound' (object, optional): If an item is found, provide its 'name' and 'description'. Omit this field if no item is found.
- For 'imagePrompt' (string): Provide a descriptive prompt for image generation relevant to the scene."
`;

exports[`server/prompts.ts generateWorldDetailsPrompt should generate the correct prompt for world details 1`] = `
"You are a world-building AI. Based on the provided adventure outline, player persona, and genre, generate detailed world information.
Adventure Title: "The Sunless Citadel"
Overall Goal: "Reclaim the Sunstone from the heart of the citadel."
Adventure Stages:
  Stage 1: "Entry" - Breach the citadel gates. (Objective: Find a way in.)
  Stage 2: "The Undercroft" - Navigate the haunted halls. (Objective: Reach the inner sanctum.)
  Stage 3: "Sanctum of Shadows" - Confront the guardian and claim the Sunstone. (Objective: Defeat the guardian.)
Player Persona: "Grim Warden" (base archetype: Brave Warrior)
Adventure Genre: Dark Fantasy

Generate rich and interconnected world details. These details should directly influence the atmosphere, potential encounters, challenges, and items within the adventure.
The player's persona (Grim Warden, archetype Brave Warrior) might have unique insights or connections to certain aspects of this world.

Provide content for the following fields:
- 'worldName': A unique and evocative name for this game world or region, fitting for a Grim Warden.
- 'genreClarification': A more specific clarification of the genre, possibly blending sub-genres (e.g., "High fantasy with elements of cosmic horror" or "Dark fantasy survival in a post-magical apocalypse").
- 'keyEnvironmentalFeatures' (array of strings): List 2-3 distinct and striking environmental features or geographical oddities. Each feature should be a string. Example values: "A constantly shifting crystal desert", "Floating islands wreathed in perpetual storms", "A forest where trees whisper prophecies, sometimes containing \\"forbidden truths\\"."
- 'dominantSocietiesOrFactions' (array of strings): Describe 1-2 major societies, factions, or sentient species. Each description is a string. e.g., "The reclusive Sky-Elves of Mount Cinder, known for their powerful elemental magic and distrust of outsiders."
- 'uniqueCreaturesOrMonsters' (array of strings): Name and briefly describe 1-2 unique creatures or monsters. Each description is a string. e.g., "Chronomites: small, insectoid creatures that can locally distort time.", "Grief-fiends: ethereal beings that feed on sorrow."
- 'magicSystemOverview': Briefly describe the nature of magic in this world. This should be a string. e.g., "Magic is a wild, untamed force drawn from the raw elements, accessible only to those with innate talent or through dangerous pacts."
- 'briefHistoryHook': A short, intriguing piece of history or lore relevant to the adventure. This should be a string. e.g., "The land is still scarred by the 'War of Whispers' a century ago, where forbidden knowledge almost unmade reality."
- 'culturalNormsOrTaboos' (array of strings): List 1-2 significant cultural norms, traditions, or taboos. Each is a string. e.g., "Offering a shard of obsidian is a sign of respect.", "Speaking the name of the last Tyrant King is forbidden and believed to bring misfortune."

Ensure all fields are filled with creative and relevant information."
`;
